src/common/lib/client/realtimepresence.ts

```diff
@@ -224,92 +201,72 @@ class RealtimePresence extends EventEmitter {
       presence.clientId = clientId;
     }
 
-    switch (channel.state) {
-      case 'attached':
-        channel.sendPresence(presence, callback);
-        break;
-      case 'attaching':
-        this.pendingPresence.push({
-          presence: presence,
-          callback: callback,
-        });
-        break;
-      case 'initialized':
-      case 'failed': {
-        /* we're not attached; therefore we let any entered status
-         * timeout by itself instead of attaching just in order to leave */
-        const err = new PartialErrorInfo('Unable to leave presence channel (incompatible state)', 90001);
-        callback?.(err);
-        break;
+    return new Promise((resolve, reject) => {
+      switch (channel.state) {
+        case 'attached':
+          channel.sendPresence(presence, (err) => (err ? reject(err) : resolve()));
+          break;
+        case 'attaching':
+          this.pendingPresence.push({
+            presence: presence,
+            callback: (err) => (err ? reject(err) : resolve()),
+          });
+          break;
+        case 'initialized':
+        case 'failed': {
+          /* we're not attached; therefore we let any entered status
+           * timeout by itself instead of attaching just in order to leave */
+          throw new PartialErrorInfo('Unable to leave presence channel (incompatible state)', 90001);
+        }
+        default:
+          throw channel.invalidStateError();
```

Comment:
should call reject with this error

Reply:
https://github.com/ably/ably-js/pull/1541#discussion_r1467782138

Reply:
https://github.com/ably/ably-js/pull/1541#discussion_r1471715976